### DISCUSSION TITLE
Using Cue as a Source Schema for Integration Framework

### DISCUSSION THREAD

**Discussion by rawkode** [cuelang/cue#482] [General]
Hi,

I'm investigating using Cue as the source schema to define the shape of various entities within a system I am building.

Using Cue, I can validate YAML and JSON documents that contain objects that are defined by the schema I provide.

However, I wish to provide integrations that allow other tools to then consume these schemas too. Initial proof of concept requires GraphQL (Schema) and Rust (Struct)

Example:

I know this is old Cue, I need to update soon.

```cue
package events
import "time"
EventType :: "Conference" | "User Group"
Event :: {
  id: =~ "^[a-z0-9_-]+$"
  name: string
  type: EventType
  dateStart:  time.Time()
  dateFinish: time.Time() & >dateStart
}
```

Should output GraphQL like such:

```graphql
enum EventType {
  Conference = "Conference"
  UserGroup = "User Group"
}

type Event {
  id: String!
  name: String!
  type: EventType!

  dateStart: DateTime!
  dateFinish: DateTime!
}

```

and Rust, like such:

```rust
use chrono::{DateTime, Utc};

pub enum EventType {
    Conference,
    UserGroup,
}

pub struct Event {
    pub id: String,
    pub name: String,
    pub r#type: EventType,
    pub date_start: DateTime<Utc>,
    pub date_end: DateTime<Utc>,
}
```

I'm currently looking into how I can use `cue/cmd` to do this, and I've received a few tips from @myitcv.

Starting this discussion to see if others are doing the same and could potentially provide some examples to speed me up.

Also to discuss these encodings being part of Cue, if others are looking for this.

**Comment by orpharion** [comment]
Not sure if you're still looking at this, but I've been writing encoders / decoders that work on either the cue Syntax tree or Value tree for python and markdown. I've done it using cue/cue and having a look at cue's own parser, it's not been too bad. I imagine that, generally, the translators for various languages will look very similar. Can you share your experience of doing it with cue/cmd?

**Comment by kumorid** [comment]
Well, I believe I am looking into a similar problem: manipulating CUE files via a gui tool.

The gui tool would be used to build the structure of elements we are defining. As our model schemas are defined in CUE, we would like to produce a representation tat can be managed by the visual tool (javascript).

I would be interested in pointers to where to start from. Ideally, there would alredy be something fully/half available.

**Comment by verdverm** [comment]
You can use the Cue `template/text` package to leverage Go templates behind the scenes.

This is essentially what my https://github.com/hofstadter-io/hof does behind the scenes using the Cue Go API. You could do the same with pure Cue too. We use the Go API because it opens up more possibilities and is more efficient when purpose built for this task.

**Reply by mpvl** [reply]
Yes, this is approach I would suggest. It may be worth building a library of these things as people make such converters. @myitcv is maintaining a repo that good be a good place for such things.

**Comment by mpvl** [comment]
In addition, we are moving much of the functionality of the `cue` command to separate "framework" packages. This will allow tool makers to create their own tools with similar functionality more easily and with more consistency.

In that vain, ik makes sense to also allow adaptors for different file types of some kind that can be added to the `load` package.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/482.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
